src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentComplexAttributeContentPass.cs

```diff
@@ -36,88 +36,86 @@ private void ProcessAttributes(TagHelperIntermediateNode node)
-            htmlNode.Children.Count > 1)
+        var removeNode = false;
+        var issueDiagnostic = false;
+
+        if (node.Children is [HtmlAttributeIntermediateNode { Children.Count: > 1 }])
         {
             // This case can be hit for a 'string' attribute
-            return false;
+            removeNode = true;
+            issueDiagnostic = true;
         }
-        else if (node.Children.Count == 1 &&
-            node.Children[0] is CSharpExpressionIntermediateNode cSharpNode &&
-            cSharpNode.Children.Count > 1)
+        else if (node.Children is [CSharpExpressionIntermediateNode { Children.Count: > 1 } cSharpNode])
         {
             // This case can be hit when the attribute has an explicit @ inside, which
             // 'escapes' any special sugar we provide for codegen.
             //
             // There's a special case here for explicit expressions. See https://github.com/aspnet/Razor/issues/2203
             // handling this case as a tactical matter since it's important for lambdas.
-            if (cSharpNode.Children.Count == 3 &&
-                cSharpNode.Children[0] is IntermediateToken token0 &&
-                cSharpNode.Children[2] is IntermediateToken token2 &&
-                token0.Content == "(" &&
-                token2.Content == ")")
+            if (cSharpNode.Children is [IntermediateToken { Content: "(" }, _, IntermediateToken { Content: ")" }])
             {
                 cSharpNode.Children.RemoveAt(2);
                 cSharpNode.Children.RemoveAt(0);
-
-                // We were able to simplify it, all good.
-                return true;
             }
-
-            return false;
+            else
+            {
+                removeNode = true;
+                issueDiagnostic = true;
+            }
         }
-        else if (node.Children.Count == 1 &&
-            node.Children[0] is CSharpCodeIntermediateNode)
+        else if (node.Children is [CSharpCodeIntermediateNode])
         {
             // This is the case when an attribute contains a code block @{ ... }
             // We don't support this.
-            return false;
+            removeNode = true;
+            issueDiagnostic = true;
+        }
+        else if (node.Children is [CSharpExpressionIntermediateNode, HtmlContentIntermediateNode { Children: [IntermediateToken { Content: "." }] }])
+        {
+            // This is the case when an attribute contains something like "@MyEnum."
+            // We simplify this to remove the "." so that tooling can provide completion on "MyEnum"
+            // in case the user is in the middle of typing
+            node.Children.RemoveAt(1);
+
+            // We still want to issue a diagnostic, even though we simplified, because ultimately
+            // we don't support this, so if the user isn't typing, we can't let this through
+            issueDiagnostic = true;
         }
```

Comment:
Could set `removeNode = true` at runtime, but with the merge of the two modes, it seemed overkill.